src/protocol/boolean/random_bits.rs

```diff
@@ -0,0 +1,129 @@
+use crate::error::BoxError;
+use crate::ff::{Field, Int};
+use crate::protocol::boolean::BitOpStep;
+use crate::protocol::modulus_conversion::convert_shares::{ConvertShares, XorShares};
+use crate::protocol::{context::ProtocolContext, RecordId};
+use crate::secret_sharing::Replicated;
+use futures::future::try_join_all;
+
+/// This protocol generates a sequence of uniformly random sharing of bits in `F_p`.
+/// Adding these 3-way secret-sharing will yield the secret `r_i ∈ {0,1}`.
+pub struct RandomBits {}
+
+impl RandomBits {
+    #[allow(dead_code)]
+    pub async fn execute<F: Field>(
+        ctx: ProtocolContext<'_, Replicated<F>, F>,
+        record_id: RecordId,
+    ) -> Result<(Vec<Replicated<F>>, Replicated<F>), BoxError> {
+        // We assume the bit length we operate in would not exceed 255
+        #[allow(clippy::cast_possible_truncation)]
+        let l = F::Integer::BITS as u8;
+
+        // Generate a replicated random field values. We'll use these fields as
+        // the source of `l`-bit long uniformly random sequence of bits.
+        let r_p: Replicated<F> = ctx.prss().generate_replicated(record_id);
+
+        // Convert each bit to secret sharings of that bit in the target field
+        let futures = (0..l).map(|i| {
+            // again, we don't expect our prime field to be > 2^64
+            #[allow(clippy::cast_possible_truncation)]
+            let xor_shares = XorShares {
+                num_bits: l,
+                packed_bits_left: r_p.left().as_u128() as u64,
+                packed_bits_right: r_p.right().as_u128() as u64,
+            };
+            let c = ctx.narrow(&BitOpStep::Step(i as usize));
+            async move {
+                ConvertShares::new(xor_shares)
+                    .execute_one_bit(c, record_id, i as u8)
+                    .await
+            }
+        });
+        let r_b = try_join_all(futures).await?;
+
+        Ok((r_b, r_p))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::RandomBits;
+    use crate::{
+        error::BoxError,
+        ff::{Field, Fp31, Fp32BitPrime},
+        protocol::{QueryId, RecordId},
+        test_fixture::{
+            make_contexts, make_world, validate_and_reconstruct, validate_and_reconstruct_xor,
+            TestWorld,
+        },
+    };
+    use futures::future::try_join_all;
+    use rand::{distributions::Standard, prelude::Distribution};
+
+    /// Take a slice of bits in `{0,1} ⊆ F_p`, and reconstruct the integer in `F_p`
+    fn bits_to_field<F: Field>(x: &[F]) -> F {
+        #[allow(clippy::cast_possible_truncation)]
+        let v = x
+            .iter()
+            .enumerate()
+            .fold(0, |acc, (i, &b)| acc + 2_u128.pow(i as u32) * b.as_u128());
+        F::from(v)
+    }
```

Comment:
Is this duplicate code? I think this appears in two places now. Can you just move this into text_fixtures and call it from both places?